### PR TITLE
Add a command to explicitly purge the metastore

### PR DESCRIPTION
**Pull request by eric-maynard**
# Description

This adds a new command, `purge`, that must be run in order to purge entities from the metastore.

Additionally, bootstrapping will now fail if entities exist in the metastore instead of purging them.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Previous behavior:

```
java -jar .app/polaris-service-1.0.0-all.jar bootstrap polaris-server.yml
. . .
(Succeeds)
```

New behavior:
```
java -jar .app/polaris-service-1.0.0-all.jar bootstrap polaris-server.yml
. . .
It appears this metastore manager has already been bootstrapped...
```

New behavior + purge:
```
java -jar .app/polaris-service-1.0.0-all.jar purge polaris-server.yml
. . .
(Succeeds)
. . .
java -jar .app/polaris-service-1.0.0-all.jar bootstrap polaris-server.yml
(Succeeds)
```

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by snazy — COMMENTED** [review]
I'm on the fence with some `--overwrite` _flag_ that purges (some) state.

`bootstrap` is a one-off operation that, I think, should fail when the system's already bootstrapped.

I think, it's better to have an explicit command or tool to purge / reset to avoid accidentally resetting principals. WDYT?

**Comment by eric-maynard** [comment]
Thanks for taking a look @snazy -- to be clear, the current behavior is that bootstrap will silently purge the state. This PR proposes a flag you have to set in order to confirm that you really want to do this. And without that flag bootstrap will fail when the system is already bootstrapped.

**Comment by snazy** [comment]
Oh - right - that "bark at you" code part is new. Can we add the "barking" part and have a "purge root-principals" command or so?

**Comment by snazy** [comment]
I feel that's a bit safer - humans tend to be lazy (and forget that they have such `--overwrite` flags set) and wonder why things break.

File: polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java [hunk 1]
```diff
@@ -143,14 +143,35 @@ public void setStorageIntegrationProvider(PolarisStorageIntegrationProvider stor
...
     // While bootstrapping we need to act as a fake privileged context since the real
     // CallContext hasn't even been resolved yet.
     PolarisCallContext polarisContext =
         new PolarisCallContext(
             sessionSupplierMap.get(realmContext.getRealmIdentifier()).get(), diagServices);
     CallContext.setCurrentContext(CallContext.of(realmContext, polarisContext));
 
+    if (!overwrite) {
+      PolarisMetaStoreManager.EntityResult preliminaryRootPrincipalLookup =
+          metaStoreManager.readEntityByName(
```
Comment (RussellSpitzer) [inline polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:158]: Will this be version safe in the future? For now I assume this is safe but in the future could readEntityByName fail for a Polaris catalog bootstrapped from an earlier version?

This seems like a good check to have for now, I'm just more worried about what happens if this call fails because the current version of Polaris just doesn't understand how to look up the Principal in the past.

File: polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java [hunk 2]
```diff
@@ -42,5 +42,5 @@ public interface MetaStoreManagerFactory extends Discoverable {
 
   void setMetricRegistry(PolarisMetricRegistry metricRegistry);
 
-  Map<String, PolarisMetaStoreManager.PrincipalSecretsResult> bootstrapRealms(List<String> realms);
```
Comment (RussellSpitzer) [inline polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java:45]: May not be an issue since this is a relatively early time in the project but we may want to get in the habit of being careful when changing method signatures. This is a public interface so we probably would want to keep
```java
bootstrapRealms(List<String> realms) {
   bootstrapRealms(realms, false)
}
```

To avoid breaking implementers

File: polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java (hunk 1 above)
Comment (eric-maynard) [reply polaris-core/src/main/java/io/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:158]: This is an important concern and it's why we don't rely on a check like this to auto-bootstrap... but it seems to me that this check is much less critical. That scenario would be very bad for other reasons, but here you would only be losing the protection of the `--overwrite` flag.

File: polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java (hunk 2 above)
Comment (eric-maynard) [reply polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java:45]: This is package-private and not public, right?

**Review by RussellSpitzer — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
@snazy I think we probably can't safely always validate (or require that an implementation can validate) that existing information is in the destination. I like this though because it is at least safer than the current behavior.

File: polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java (hunk 2 above)
Comment (snazy) [reply polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java:45]: It's `public` (Java interface, not a Java abstract class).

**Comment by snazy** [comment]
> I think we probably can't safely always validate (or require that an implementation can validate) that existing information is in the destination. I like this though because it is at least safer than the current behavior.

Sure, having [this check](https://github.com/polaris-catalog/polaris/pull/59/files#diff-bec903bf82bc195032d50b148f8e1c413cf20be1245ad45901219cbf138b0035R164) is definitely a good thing. What I'm thinking of is:
* not having an `--overwrite` flag, but let `bootstrap` always fail with that check but
* have a new command to explicitly purge the principal

Thinking of humans leaving the `--overwrite` in for all deployments - (and complain that authn no longer works after a restart).

**Comment by eric-maynard** [comment]
I don't think we want to break bootstrap into purge + bootstrap. Not all metastore managers will necessarily purge on bootstrap. We want to just gate bootstrapping behind an additional flag on a best-effort basis in the case that we detect bootstrapping may have already been performed.

`--overwrite` doesn't appear in the docs and only appears in this error message, so for now it's hard to imagine someone always leaving `--overwrite` on their bootstrap command without knowing what it does. And if we can imagine such a state, it's not such a stretch to also imagine that they leave the `purge` command in there too.

**Comment by sfc-gh-mcollado** [comment]
IDK, I like the option to `purge` because I feel it is very explicit and we should expect the implementation to be thorough. `bootstrap` is necessarily an unauthenticated command, whereas the `purge` operation ought to require root credentials (there's a question of how does someone bootstrap if they lost their root credentials, at which point I think they should require direct database access). 

If `bootstrap --overrwrite` merely focuses on replacing the root principal/secrets, there's a risk that someone could `bootstrap` an existing installation but not delete the pre-existing data. In my mind, `purge` means _delete all the things_, so `purge` then `bootstrap` should result in the expected behavior.

**Comment by collado-mike** [comment]
IDK, I like the option to purge because I feel it is very explicit and we should expect the implementation to be thorough. bootstrap is necessarily an unauthenticated command, whereas the purge operation ought to require root credentials (there's a question of how does someone bootstrap if they lost their root credentials, at which point I think they should require direct database access).

If bootstrap --overrwrite merely focuses on replacing the root principal/secrets, there's a risk that someone could bootstrap an existing installation but not delete the pre-existing data. In my mind, purge means delete all the things, so purge then bootstrap should result in the expected behavior.

File: polaris-server.yml [hunk 3]
```diff
@@ -88,9 +88,9 @@ defaultRealms:
   - default-realm
 
 metaStoreManager:
-  type: in-memory
-  # type: eclipse-link # uncomment to use eclipse-link as metastore
-  # persistence-unit: polaris-dev
+  # type: in-memory
```
Comment (snazy) [inline polaris-server.yml:91]: Can you revert this change?

**Review by snazy — COMMENTED** [review]
One nit, but overall LGTM.
Deferring review to @collado-mike, because this related code base a bit better ;)

File: polaris-core/src/main/java/io/polaris/core/persistence/PolarisMetaStoreManager.java [hunk 4]
```diff
@@ -176,18 +176,28 @@ public boolean alreadyExists() {
...
-   * Bootstrap the Polaris service, will remove ALL existing persisted entities, then will create
-   * the root catalog, root principal and associated service admin role.
+   * Bootstrap the Polaris service, creating the root catalog, root principal, and associated
+   * service admin role. Will fail if the service has already been bootstrapped.
+   *
+   * @param callCtx call context
+   * @return always success or unexpected error
+   */
+  @NotNull
+  BaseResult bootstrapPolarisService(@NotNull PolarisCallContext callCtx);
```
Comment (collado-mike) [inline polaris-core/src/main/java/io/polaris/core/persistence/PolarisMetaStoreManager.java:186]: Currently, the behavior is to throw an `IllegalArgumentException` if the realm cannot be bootstrapped. It seems for consistency with the other APIs, it should return the `BaseResult` with `success=false` rather than throwing the exception. Let the `bootstrap` command deal with the failure state

File: polaris-core/src/main/java/io/polaris/core/persistence/PolarisMetaStoreManagerImpl.java [hunk 5]
```diff
@@ -706,6 +703,18 @@ private void bootstrapPolarisService(
     return new BaseResult(ReturnStatus.SUCCESS);
   }
 
+  @Override
+  public @NotNull BaseResult purge(@NotNull PolarisCallContext callCtx) {
+    // get meta store we should be using
+    PolarisMetaStoreSession ms = callCtx.getMetaStore();
+
+    // run operation in a read/write transaction
+    ms.runActionInTransaction(callCtx, () -> ms.deleteAll(callCtx));
```
Comment (collado-mike) [inline polaris-core/src/main/java/io/polaris/core/persistence/PolarisMetaStoreManagerImpl.java:712]: Stick a `LOGGER.warn` statement in there :)

File: polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java [hunk 6]
```diff
@@ -56,5 +59,7 @@ protected void run(
     }
 
     metaStoreManagerFactory.bootstrapRealms(configuration.getDefaultRealms());
```
Comment (collado-mike) [inline polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java:61]: Since the method has a return value, check its status here.

File: polaris-service/src/main/java/io/polaris/service/PurgeRealmsCommand.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import io.polaris.service.config.PolarisApplicationConfig;
+import io.polaris.service.config.RealmEntityManagerFactory;
+import io.polaris.service.context.CallContextResolver;
+import net.sourceforge.argparse4j.inf.Namespace;
+import org.slf4j.Logger;
+
+/**
+ * Command for purging root level service principals for a realm. This command will invoke a default
+ * implementation which generates random user id and secret. These credentials will be printed out
+ * to the log and standard output (stdout).
```
Comment (collado-mike) [inline polaris-service/src/main/java/io/polaris/service/PurgeRealmsCommand.java:32]: This is false, right? It just purges the data and returns nothing?

**Review by collado-mike — CHANGES_REQUESTED** [review]

File: polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java [hunk 8]
```diff
@@ -43,4 +43,6 @@ public interface MetaStoreManagerFactory extends Discoverable {
   void setMetricRegistry(PolarisMetricRegistry metricRegistry);
 
   Map<String, PolarisMetaStoreManager.PrincipalSecretsResult> bootstrapRealms(List<String> realms);
+
+  void purgeRealms(List<String> realms);
```
[the remark this answers is not in the document]
Comment (eric-maynard) [reply polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java:47]: Good call -- note that the existing methods don't have one, we probably need to take a pass and add a lot of these

[the remark this answers is not in the document]
Comment (collado-mike) [reply polaris-core/src/main/java/io/polaris/core/persistence/MetaStoreManagerFactory.java:48]: yeah, I know... boy scout rule :)

File: polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java [hunk 9]
```diff
@@ -55,6 +60,27 @@ protected void run(
...
+    // Execute the bootstrap
+    Map<String, PolarisMetaStoreManager.PrincipalSecretsResult> results =
+        metaStoreManagerFactory.bootstrapRealms(configuration.getDefaultRealms());
+
+    // Log any errors:
+    boolean success = true;
+    for (Map.Entry<String, PolarisMetaStoreManager.PrincipalSecretsResult> result :
+        results.entrySet()) {
+      if (!result.getValue().isSuccess()) {
+        LOGGER.warn(
```
Comment (collado-mike) [inline polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java:72]: I'd make this an `error` since this is explicitly the `bootstrap` command

**Review by collado-mike — DISMISSED** [review]

**Comment by MonkeyCanCode** [comment]
@flyrain can we have this merge if not other blockers so the bootstrap won't wipe the backend if accidently got ran more than once. Then I can test if anything additional will be needed on the helm side as well to handle those.

File: polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java [hunk 10]
```diff
@@ -55,6 +60,27 @@ protected void run(
       csa.setConfigurationStore(configurationStore);
     }
 
-    metaStoreManagerFactory.bootstrapRealms(configuration.getDefaultRealms());
+    // Execute the bootstrap
+    Map<String, PolarisMetaStoreManager.PrincipalSecretsResult> results =
+        metaStoreManagerFactory.bootstrapRealms(configuration.getDefaultRealms());
```
Comment (flyrain) [inline polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java:65]: Nit: we can use `var` to simplify a bit.

File: polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java [hunk 11]
```diff
@@ -55,6 +60,27 @@ protected void run(
...
+
+    // Log any errors:
+    boolean success = true;
+    for (Map.Entry<String, PolarisMetaStoreManager.PrincipalSecretsResult> result :
+        results.entrySet()) {
```
Comment (flyrain) [inline polaris-service/src/main/java/io/polaris/service/BootstrapRealmsCommand.java:70]: Nit: We can use `var` to simplify and keep them in one line

**Review by flyrain — APPROVED** [review]
LGTM with minor comments.

**Comment by eric-maynard** [comment]
I haven’t seen any use of `var` outside of tests in this project; might
there be compatibility concerns? Is this merge blocking?

**Comment by flyrain** [comment]
I think it's fine to use `var` as Polaris requires jdk 11 for core and jdk21 for others. `var` was introduced in jdk 10. To be clear, they are not blockers.

**Comment by flyrain** [comment]
Thanks @eric-maynard for the change. Thanks @collado-mike @snazy @RussellSpitzer @MonkeyCanCode for the review.
